cmd/m3db-operator/main.go

```diff
@@ -168,23 +172,42 @@ func main() {
 		logger.Fatal("failed to create k8sclient", zap.Error(err))
 	}
 
-	stopCh := make(chan struct{})
-	var kubeInformerFactory kubeinformers.SharedInformerFactory
-	if _namespace == "all" {
-		kubeInformerFactory = kubeinformers.NewSharedInformerFactory(kubeClient, _informerSyncDuration)
-	} else {
-		kubeInformerFactory = kubeinformers.NewSharedInformerFactoryWithOptions(kubeClient, _informerSyncDuration, kubeinformers.WithNamespace(_namespace))
+	labelReq, err := klabels.NewRequirement(labels.App, selection.Exists, nil)
+	if err != nil {
+		logger.Fatal("failed to construct selector requirement", zap.Error(err))
 	}
+	tweakOpts := kubeinformers.WithTweakListOptions(func(opts *metav1.ListOptions) {
+		// PodSelector provides a selector for listing only M3DB pods to annotate.
+		podSelector := klabels.NewSelector().Add(*labelReq)
+		opts.LabelSelector = podSelector.String()
+	})
 
-	nodeLister := kubeInformerFactory.Core().V1().Nodes().Lister()
+	var (
+		m3dbClusterInformerFactory informers.SharedInformerFactory
+		// filteredInformerFactory is used to create the pod and statefulset
+		// informers. We know that every pod and StatefulSet created by the operator
+		// will have its hardcoded labels. The pod annotation and statefulset loop
+		// may slow down in clusters with many pods or statefulsets if it must
+		// process every objects in the cluster. The filter only lists objects that
+		// have labels we expect.
+		filteredInformerFactory kubeinformers.SharedInformerFactory
+		// The kube informer factory is still required to list nodes, as they don't
+		// have our label.
+		kubeInformerFactory kubeinformers.SharedInformerFactory
```

Comment:
Do we need a shared informer factory for the node lister?

I checked where `nodeLister` is used and it's only with `.Get(..)` (i.e. lookup a specific host) so would it be possible to just use the kube client for that rather than an informer factory which can be affected by the same issue we're solving here?

Reply:
If we use the informer then we'll be notified of updates, and our `Get(...)` calls will just read from the cache if it's fresh. If we were to use the client directly it would put slightly more load on the API server.